### PR TITLE
Add initial source documentation for Platypus kernels and boundary conditions

### PR DESCRIPTION
**Pull request by alexanderianblair**
Adds initial source documentation for kernels and boundary conditions.

### CODE REVIEW

File: doc/content/source/kernels/MFEMCurlCurlKernel.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+!syntax description /Kernels/MFEMCurlCurlKernel
+
+## Overview
+
+Adds the domain integrator for integrating the bilinear form
+
+!equation
+(k\vec\nabla \times \vec u, \vec\nabla \times \vec v)_\Omega \,\,\, \forall \vec v \in V
+
+where $\vec u, \vec v \in H(\mathrm{curl})$.
```
Comment (cmacmackin) [inline doc/content/source/kernels/MFEMCurlCurlKernel.md:14]: Add a definition for k.

File: doc/content/source/kernels/MFEMDiffusionKernel.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+## Summary
+
+!syntax description /Kernels/MFEMDiffusionKernel
+
+## Overview
+
+Adds the domain integrator for integrating the bilinear form
+
+!equation
+(k\vec\nabla u, \vec\nabla v)_\Omega \,\,\, \forall v \in V
```
Comment (cmacmackin) [inline doc/content/source/kernels/MFEMDiffusionKernel.md:12]: Add a definition for k

File: doc/content/source/kernels/MFEMMassKernel.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+## Summary
+
+!syntax description /Kernels/MFEMMassKernel
+
+## Overview
+
+Adds the domain integrator for integrating the bilinear form
+
+!equation
+(k u, v)_\Omega \,\,\, \forall v \in V
```
Comment (cmacmackin) [inline doc/content/source/kernels/MFEMMassKernel.md:12]: Add description of k?

File: doc/content/source/kernels/MFEMMixedVectorGradientKernel.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+## Summary
+
+!syntax description /Kernels/MFEMMixedVectorGradientKernel
+
+## Overview
+
+Adds the domain integrator for integrating the mixed bilinear form
+
+!equation
+(k \vec \nabla u, \vec v)_\Omega \,\,\, \forall \vec v \in V
```
Comment (cmacmackin) [inline doc/content/source/kernels/MFEMMixedVectorGradientKernel.md:12]: Add description of k?

File: doc/content/source/kernels/MFEMTimeDerivativeMassKernel.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+## Summary
+
+!syntax description /Kernels/MFEMTimeDerivativeMassKernel
+
+## Overview
+
+Adds the domain integrator for integrating the bilinear form
+
+!equation
+(k \dot{u}, v)_\Omega \,\,\, \forall v \in V
```
Comment (cmacmackin) [inline doc/content/source/kernels/MFEMTimeDerivativeMassKernel.md:12]: Add description of k?

File: doc/content/source/kernels/MFEMVectorFEDomainLFKernel.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
+# MFEMVectorFEDomainLFKernel
+
+!syntax description /Kernels/MFEMVectorFEDomainLFKernel
+
+## Overview
+
+Adds the domain integrator for integrating the linear form
+
+!equation
+(\vec f, \vec v)_\Omega \,\,\, \forall \vec v \in V
```
Comment (cmacmackin) [inline doc/content/source/kernels/MFEMVectorFEDomainLFKernel.md:10]: Add description of v

File: doc/content/source/kernels/MFEMVectorFEMassKernel.md [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+## Summary
+
+!syntax description /Kernels/MFEMVectorFEMassKernel
+
+## Overview
+
+Adds the domain integrator for integrating the bilinear form
+
+!equation
+(k \vec u, \vec v)_\Omega \,\,\, \forall \vec v \in V
```
Comment (cmacmackin) [inline doc/content/source/kernels/MFEMVectorFEMassKernel.md:12]: Add description of k?

File: doc/content/source/kernels/MFEMVectorFEWeakDivergenceKernel.md [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+!syntax description /Kernels/MFEMVectorFEWeakDivergenceKernel
+
+## Overview
+
+Adds the domain integrator for integrating the mixed bilinear form
+
+!equation
+-(k \vec u, \vec\nabla v)_\Omega \,\,\, \forall v \in V
+
+where $\vec u \in H(\mathrm{curl})$ or $H(\mathrm{div})$ and
```
Comment (cmacmackin) [inline doc/content/source/kernels/MFEMVectorFEWeakDivergenceKernel.md:14]: Add description of k?

**Review by cmacmackin — CHANGES_REQUESTED** [review]
There are a few bits where I think some additional description could be useful.

**Comment by cmacmackin** [comment]
I'm guessing the CI failure is because you're trying to link to a page that got split into another PR?

**Comment by alexanderianblair** [comment]
Yes - the CI failure is indeed due to a reference to a file in another PR (no perfect way to do this, unfortunately). Any issues on this front should be caught in the feature branch when the split PRs are merged into it.

File: doc/content/source/bcs/MFEMBoundaryCondition.md [hunk 9]
```diff
@@ -0,0 +1,23 @@
+# MFEMBoundaryCondition
+
+## Summary
+
+Base class for MFEM boundary condition objects.
+
+## Overview
+
+`MFEMBoundaryCondition` represent boundary conditions applied to the MFEM problem, applied on the
```
Comment (nmnobre) [inline doc/content/source/bcs/MFEMBoundaryCondition.md:9]:
```suggestion
`MFEMBoundaryCondition` represents boundary conditions applied to the MFEM problem, applied on the
```

File: doc/content/source/bcs/MFEMBoundaryCondition.md [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+labelled according to the test variable name returned from `getTestVariableName()`.
+
+`MFEMBoundaryCondition` objects come in two varieties, depending on which child class they inherit
+from:
+
+- Essential boundary conditions, inheriting from `MFEMEssentialBC`, for the application of
+  Dirichlet-like BCs removing degrees of freedom from the problem at the boundary;
+
+- Integrated boundary conditions, inheriting from `MFEMIntegratedBC`, which apply one or more
+  boundary integrators to the weak form on the specified boundary surfaces.
```
Comment (nmnobre) [inline doc/content/source/bcs/MFEMBoundaryCondition.md:22]: "boundary surfaces" or just "boundaries"? Do we support 2d problems?

File: doc/content/source/bcs/MFEMConvectiveHeatFluxBC.md [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+## Overview
+
+Adds the boundary integrator for integrating the bilinear form
+
+!equation
+\lambda(u-u_0, v)_{\partial\Omega} \,\,\, \forall v \in V
+
+where $u, v \in H^1$ are the test and trial variables, and $\lambda, u_0$ are scalar functions on the surface independent of $u$.
+
+This boundary condition is particularly useful for use in thermal problems, where it can be used to
```
Comment (nmnobre) [inline doc/content/source/bcs/MFEMConvectiveHeatFluxBC.md:16]:
```suggestion
This boundary condition is particularly useful for thermal problems, where it can be used to
```

File: doc/content/source/bcs/MFEMConvectiveHeatFluxBC.md [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+represent a heat transfer boundary condition
+
+!equation
+h(T-T_{\inf}, T')_{\partial\Omega} \,\,\, \forall T' \in V
+
+by identifying $\lambda=h$ as the heat transfer coefficient on the boundary, $u=T$ as the trial variable
+for the temperature, $v=T'$ is the test variable, and $u_0=T_{inf}$ as the equilibrium temperature far from the boundary.
```
Comment (nmnobre) [inline doc/content/source/bcs/MFEMConvectiveHeatFluxBC.md:23]:
```suggestion
for the temperature, $v=T'$ as the test variable, and $u_0=T_{inf}$ as the equilibrium temperature far from the boundary.
```

File: doc/content/source/bcs/MFEMVectorNormalIntegratedBC.md [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+!syntax description /BCs/MFEMVectorNormalIntegratedBC
+
+## Overview
+
+Adds the boundary integrator for integrating the linear form
+
+!equation
+(\vec f \cdot \vec n, v)_{\partial\Omega} \,\,\, \forall v \in V
+
+where $v \in H^1$ and $\vec n$ is the outward facing unit normal vector on the surface.
```
Comment (nmnobre) [inline doc/content/source/bcs/MFEMVectorNormalIntegratedBC.md:14]: Do we want to use $\vec n$ or $\hat n$ throughout for unit normals?

File: doc/content/source/bcs/MFEMBoundaryCondition.md [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+
+Base class for MFEM boundary condition objects.
+
+## Overview
+
+`MFEMBoundaryCondition` represent boundary conditions applied to the MFEM problem, applied on the
+user-provided list of boundaries.
+
+Similar to `MFEMKernel`, the boundary condition is applied to the weak form equation that is
+labelled according to the test variable name returned from `getTestVariableName()`.
```
Comment (nmnobre) [inline doc/content/source/bcs/MFEMBoundaryCondition.md:13]:
```suggestion
labeled according to the test variable name returned from `getTestVariableName()`.
```

File: doc/content/source/kernels/MFEMKernel.md [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+Base class for MFEM kernels applied to the weak form being solved.
+
+## Overview
+
+MFEM kernels are responsible for providing domain integrators (inheriting from
+`mfem::BilinearFormIntegrator` or `mfem::LinearFormIntegrator` to add to the weak form of the FE
+problem accumulated in [EquationSystem](source/equation_systems/equation_system.md), along with any
+required marker arrays to restrict the integrator(s) to subdomains.
+
+The integrator is applied to the weak form equation that is labelled according to the test variable
```
Comment (nmnobre) [inline doc/content/source/kernels/MFEMKernel.md:14]:
```suggestion
The integrator is applied to the weak form equation that is labeled according to the test variable
```

File: doc/content/source/kernels/MFEMKernel.md [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+name of the kernel returned from `getTestVariableName()`. In the case of bilinear (or nonlinear)
+forms, the trial variable that the integrator acts on is the variable returned from
+`getTrialVariableName()`. For consistency with MOOSE, and for the ease of associating weak form
+equations (labelled by test variable) with the trial variable solved using them, the set of test
```
Comment (nmnobre) [inline doc/content/source/kernels/MFEMKernel.md:18]:
```suggestion
equations (labeled by test variable) with the trial variable solved using them, the set of test
```

File: doc/content/source/bcs/MFEMBoundaryCondition.md [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+
+`MFEMBoundaryCondition` objects come in two varieties, depending on which child class they inherit
+from:
+
+- Essential boundary conditions, inheriting from `MFEMEssentialBC`, for the application of
```
[the remark this answers is not in the document]
Comment (nmnobre) [reply doc/content/source/bcs/MFEMBoundaryCondition.md:18]: Do we have any docs for those two though?

File: doc/content/source/kernels/MFEMDiffusionKernel.md [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+
+where $u, v \in H^1$.
+
+This term arises from the weak form of the Laplacian operator
+
+!equation
+-k \nabla^2 u
```
Comment (nmnobre) [inline doc/content/source/kernels/MFEMDiffusionKernel.md:19]:
```suggestion
- \vec\nabla \cdot \left( k \vec\nabla u \right)
```

File: src/kernels/MFEMDiffusionKernel.C [hunk 19]
```diff
@@ -7,14 +7,13 @@ InputParameters
...
 {
   InputParameters params = MFEMKernel::validParams();
-  params.addClassDescription(
-      "The Laplacian operator ($-k\\nabla \\cdot \\nabla u$), with the weak "
-      "form of $ (k\\nabla \\phi_i, \\nabla u_h), to be added to an MFEM problem");
-
+  params.addClassDescription("Adds the domain integrator to an MFEM problem for the bilinear form "
+                             "$(k\\vec\\nabla u, \\vec\\nabla v)_\\Omega$ "
+                             "arising from the weak form of the Laplacian operator "
+                             "$-k\\nabla^2 u$.");
```
Comment (nmnobre) [inline src/kernels/MFEMDiffusionKernel.C:13]:
```suggestion
                             "$- \\vec\\nabla \\cdot \\left( k \\vec \\nabla u \\right)$.");
```

File: src/kernels/MFEMDiffusionKernel.C [hunk 20]
```diff
@@ -7,14 +7,13 @@ InputParameters
...
   params.addParam<std::string>("coefficient",
                                "Name of property for diffusion coefficient k to multiply "
```
Comment (nmnobre) [inline src/kernels/MFEMDiffusionKernel.C:15]:
```suggestion
                               "Name of property for diffusion coefficient k."
```

File: src/kernels/MFEMDiffusionKernel.C [hunk 21]
```diff
@@ -7,14 +7,13 @@ InputParameters
...
-                               "the Laplacian by");
-
+                               "the Laplacian by.");
```
Comment (nmnobre) [inline src/kernels/MFEMDiffusionKernel.C:16]:
```suggestion
```

**Review by nmnobre — CHANGES_REQUESTED** [review]

File: doc/content/source/bcs/MFEMBoundaryCondition.md (hunk 10 above)
Comment (alexanderianblair) [reply doc/content/source/bcs/MFEMBoundaryCondition.md:22]: "boundaries" is preferred - we don't yet, but there's no reason why we wouldn't in future

Comment (nmnobre) [reply doc/content/source/bcs/MFEMBoundaryCondition.md:22]:
```suggestion
  boundary integrators to the weak form on the specified boundaries.
```

File: doc/content/source/bcs/MFEMConvectiveHeatFluxBC.md [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+!syntax description /BCs/MFEMConvectiveHeatFluxBC
+
+## Overview
+
+Adds the boundary integrator for integrating the bilinear form
+
+!equation
+\lambda(u-u_0, v)_{\partial\Omega} \,\,\, \forall v \in V
+
+where $u, v \in H^1$ are the test and trial variables, and $\lambda, u_0$ are scalar functions on the surface independent of $u$.
```
Comment (nmnobre) [inline doc/content/source/bcs/MFEMConvectiveHeatFluxBC.md:14]:
```suggestion
where $u, v \in H^1$ are the test and trial variables, and $\lambda, u_0$ are scalar functions on the boundary independent of $u$.
```

File: doc/content/source/bcs/MFEMVectorNormalIntegratedBC.md (hunk 13 above)
Comment (nmnobre) [inline doc/content/source/bcs/MFEMVectorNormalIntegratedBC.md:14]:
```suggestion
where $v \in H^1$ and $\vec n$ is the outward facing unit normal vector on the boundary.
```

Comment (alexanderianblair) [reply doc/content/source/bcs/MFEMVectorNormalIntegratedBC.md:14]: I prefer $\hat n$, as it's clearer that it's a unit normal vector rather than just any normal vector - I'll update this

File: doc/content/source/bcs/MFEMConvectiveHeatFluxBC.md [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+
+This boundary condition is particularly useful for use in thermal problems, where it can be used to
+represent a heat transfer boundary condition
+
+!equation
+h(T-T_{\inf}, T')_{\partial\Omega} \,\,\, \forall T' \in V
+
+by identifying $\lambda=h$ as the heat transfer coefficient on the boundary, $u$ as the trial variable
```
[the remark this answers is not in the document]
Comment (alexanderianblair) [reply doc/content/source/bcs/MFEMConvectiveHeatFluxBC.md]: Updated

File: doc/content/source/kernels/MFEMCurlCurlKernel.md (hunk 1 above)
Comment (alexanderianblair) [reply doc/content/source/kernels/MFEMCurlCurlKernel.md]: Added

File: doc/content/source/kernels/MFEMMassKernel.md (hunk 3 above)
Comment (alexanderianblair) [reply doc/content/source/kernels/MFEMMassKernel.md:12]: Added

File: doc/content/source/kernels/MFEMDiffusionKernel.md (hunk 2 above)
Comment (alexanderianblair) [reply doc/content/source/kernels/MFEMDiffusionKernel.md:12]: Added

File: doc/content/source/kernels/MFEMMixedVectorGradientKernel.md (hunk 4 above)
Comment (alexanderianblair) [reply doc/content/source/kernels/MFEMMixedVectorGradientKernel.md:12]: Added

File: doc/content/source/kernels/MFEMTimeDerivativeMassKernel.md (hunk 5 above)
Comment (alexanderianblair) [reply doc/content/source/kernels/MFEMTimeDerivativeMassKernel.md:12]: Added

File: doc/content/source/kernels/MFEMVectorFEDomainLFKernel.md (hunk 6 above)
Comment (alexanderianblair) [reply doc/content/source/kernels/MFEMVectorFEDomainLFKernel.md:10]: Added, along with a description of $\vec f$

File: doc/content/source/kernels/MFEMVectorFEMassKernel.md (hunk 7 above)
Comment (alexanderianblair) [reply doc/content/source/kernels/MFEMVectorFEMassKernel.md:12]: Added

File: doc/content/source/kernels/MFEMVectorFEWeakDivergenceKernel.md (hunk 8 above)
Comment (alexanderianblair) [reply doc/content/source/kernels/MFEMVectorFEWeakDivergenceKernel.md]: Added

File: doc/content/source/bcs/MFEMVectorNormalIntegratedBC.md (hunk 13 above)
Comment (alexanderianblair) [reply doc/content/source/bcs/MFEMVectorNormalIntegratedBC.md]: Updated, both here and in the source file

File: doc/content/source/bcs/MFEMVectorDirichletBC.md [hunk 24]
```diff
@@ -0,0 +1,24 @@
...
+!syntax description /BCs/MFEMVectorDirichletBC
+
+## Overview
+
+Boundary condition for enforcing an essential (Dirichlet) boundary condition on a vector variable on the
+boundary.
+
+The components of the vector variable on which the essential boundary condition are applied are
+controlled by the `_boundary_apply_type` parameter. Currently, only the tangential components of the
+vector variable on the boundary are fixed to the values of the input vector coefficient.
```
[the remark this answers is not in the document]
Comment (alexanderianblair) [reply doc/content/source/bcs/MFEMVectorDirichletBC.md]: Agreed - `MFEMVectorDirichletBC` has also been split out in #41 for different apply types, which should be documented separately. I shall create a new PR for documentation for new source files that have been added over the course of the review period, to keep things manageable!

File: doc/content/source/bcs/MFEMVectorNormalIntegratedBC.md (hunk 13 above)
Comment (nmnobre) [reply doc/content/source/bcs/MFEMVectorNormalIntegratedBC.md]: Thanks for doing this Alex, sorry if I stepped on your toes when I changed the "surfaces" wording. I didn't completely understand if you wanted me to do it or not. :)

File: doc/content/source/bcs/MFEMBoundaryCondition.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (alexanderianblair) [reply doc/content/source/bcs/MFEMBoundaryCondition.md]: We do now! And I've added the links to them in `MFEMBoundaryCondition.md`

**Review by cmacmackin — APPROVED** [review]
